pkg/kubelet/cm/devicemanager/manager_test.go

```diff
@@ -434,6 +434,106 @@ func TestUpdateCapacityAllocatable(t *testing.T) {
+	as.Nil(err)
+
+	resource1Devs := []pluginapi.Device{
+		{ID: "R1Device1", Health: pluginapi.Healthy},
+		{ID: "R1Device2", Health: pluginapi.Healthy},
+		{ID: "R1Device3", Health: pluginapi.Unhealthy},
+	}
+	resourceName1 := "domain1.com/resource1"
+	e1 := &endpointImpl{}
+	testManager.endpoints[resourceName1] = endpointInfo{e: e1, opts: nil}
+	testManager.genericDeviceUpdateCallback(resourceName1, resource1Devs)
+
+	resource2Devs := []pluginapi.Device{
+		{ID: "R2Device1", Health: pluginapi.Healthy},
+	}
+	resourceName2 := "other.domain2.org/resource2"
+	e2 := &endpointImpl{}
+	testManager.endpoints[resourceName2] = endpointInfo{e: e2, opts: nil}
+	testManager.genericDeviceUpdateCallback(resourceName2, resource2Devs)
+
+	allocatableDevs := testManager.GetAllocatableDevices()
+	as.Equal(2, len(allocatableDevs))
+
+	devInstances1, ok := allocatableDevs[resourceName1]
+	as.True(ok)
+	as.Equal(2, len(devInstances1))
+	for _, deviceID := range []string{"R1Device1", "R1Device2"} {
+		_, ok := devInstances1[deviceID]
+		as.True(ok)
+	}
+
+	devInstances2, ok := allocatableDevs[resourceName2]
+	as.True(ok)
+	as.Equal(1, len(devInstances2))
+	_, ok = devInstances2["R2Device1"]
+	as.True(ok)
+
+}
+
+func TestGetAllocatableDevicesHealthTransition(t *testing.T) {
+	socketDir, socketName, _, err := tmpSocketDir()
+	topologyStore := topologymanager.NewFakeManager()
+	require.NoError(t, err)
+	defer os.RemoveAll(socketDir)
+	testManager, err := newManagerImpl(socketName, nil, topologyStore)
+	as := assert.New(t)
+	as.NotNil(testManager)
+	as.Nil(err)
+
+	resource1Devs := []pluginapi.Device{
+		{ID: "R1Device1", Health: pluginapi.Healthy},
+		{ID: "R1Device2", Health: pluginapi.Healthy},
+		{ID: "R1Device3", Health: pluginapi.Unhealthy},
+	}
+
+	// Adds three devices for resource1, two healthy and one unhealthy.
+	// Expects allocatable devices for resource1 to be 2.
+	resourceName1 := "domain1.com/resource1"
+	e1 := &endpointImpl{}
+	testManager.endpoints[resourceName1] = endpointInfo{e: e1, opts: nil}
+
+	testManager.genericDeviceUpdateCallback(resourceName1, resource1Devs)
+
+	allocatableDevs := testManager.GetAllocatableDevices()
+	as.Equal(1, len(allocatableDevs))
+	devInstances, ok := allocatableDevs[resourceName1]
+	as.True(ok)
+	as.Equal(2, len(devInstances))
+	for _, deviceID := range []string{"R1Device1", "R1Device2"} {
+		_, ok := devInstances[deviceID]
+		as.True(ok)
+	}
+
+	// Unhealthy device becomes healthy
+	resource1Devs = []pluginapi.Device{
+		{ID: "R1Device1", Health: pluginapi.Healthy},
+		{ID: "R1Device2", Health: pluginapi.Healthy},
+		{ID: "R1Device3", Health: pluginapi.Healthy},
+	}
+	testManager.genericDeviceUpdateCallback(resourceName1, resource1Devs)
+
+	allocatableDevs = testManager.GetAllocatableDevices()
+	as.Equal(1, len(allocatableDevs))
+	devInstances, ok = allocatableDevs[resourceName1]
+	as.True(ok)
+	as.Equal(3, len(devInstances))
+	for _, deviceID := range []string{"R1Device1", "R1Device2", "R1Device3"} {
```

Comment:
nit: should we factorize these checks (len and members) into a func?

Reply:
this could be a good idea, yes. On it.